lib/linked_list.rb

```diff
@@ -12,152 +12,330 @@ def initialize(value, next_node = nil)
+  #   end
+
+  #   previous = current
+  #   current = current.next
+  # end
+  # end
 
-    ## Advanced Exercises
-    # returns the value at the middle element in the singly linked list
-    # Time Complexity:  
-    # Space Complexity
-    def find_middle_value
-      raise NotImplementedError
+  # method to reverse the singly linked list
+  # note: the nodes should be moved and not just the values in the nodes
+  # Time Complexity: O(n) where n is the length of the linked list
+  # Space Complexity: O(1)
+  def reverse
+    current = @head
+    last_node = nil
+    next_node = current.next
+
+    if current == nil
+      return nil
+    end
+
+    while current.next != nil
+      current.next = last_node
+      last_node = current
+      current = next_node
+      next_node = current.next
     end
 
-    # find the nth node from the end and return its value
-    # assume indexing starts at 0 while counting to n
-    # Time Complexity:  
-    # Space Complexity
-    def find_nth_from_end(n)
-      raise NotImplementedError
+    current.next = last_node
+    @head = current
+    return @head.data
+  end
+
+  ## Advanced Exercises
+  # returns the value at the middle element in the singly linked list
+  # Time Complexity: O(n) where n is the length of the linked list
+  # Space Complexity: O(1)
+  def find_middle_value
+    slow = @head
+    fast = @head
+
+    if @head == nil
+      return nil
     end
 
-    # checks if the linked list has a cycle. A cycle exists if any node in the
-    # linked list links to a node already visited.
-    # returns true if a cycle is found, false otherwise.
-    # Time Complexity:  
-    # Space Complexity
-    def has_cycle
-      raise NotImplementedError
+    while fast != nil && fast.next != nil
+      fast = fast.next.next
+      slow = slow.next
     end
+    return slow
+  end
+
+  # find the nth node from the end and return its value
+  # assume indexing starts at 0 while counting to n
+  # Time Complexity: O(n) where n is the length of the linked list
+  # Space Complexity: O(1)
+  def find_nth_from_end(n)
+    return if self.length <= n
 
+    i = self.length - n - 1
+    current = @head
 
-    # Additional Exercises 
-    # returns the value in the first node
-    # returns nil if the list is empty
-    # Time Complexity:  
-    # Space Complexity
-    def get_first
-      raise NotImplementedError
+    i.times do
+      current = current.next
     end
 
-    # method that inserts a given value as a new last node in the linked list
-    # Time Complexity:  
-    # Space Complexity
-    def add_last(value)
-      raise NotImplementedError
+    return current.data
+  end
+
+  # checks if the linked list has a cycle. A cycle exists if any node in the
+  # linked list links to a node already visited.
+  # returns true if a cycle is found, false otherwise.
+  # Time Complexity: O(n) where n is the length of the linked list
+  # Space Complexity: O(1)

```

Comment:
Space complexity of O(1)?  Really?  You're making an array `visited` which will get as large as the list.  

You are also using `visited.include?(current)` which will do a loop through the array to find current.  So you have a loop in a loop or an O(n<sup>2</sup>) algorithm.